include/libpmemobj++/array.hpp

```diff
@@ -0,0 +1,724 @@
+	front()
+	{
+		detail::conditional_add_to_tx(&_data[0]);
+		return _data[0];
+	}
+
+	/**
+	 *  Access the last element and add this element to a transaction.
+	 *
+	 * @throw transaction_error when adding the object to the
+	 *		transaction failed.
+	 */
+	reference
+	back()
+	{
+		detail::conditional_add_to_tx(&_data[size() - 1]);
+		return _data[size() - 1];
+	}
+
+	/**
+	 *  Access the first element.
+	 */
+	const_reference
+	front() const
+	{
+		return _data[0];
+	}
+
+	/**
+	 *  Access the last element.
+	 */
+	const_reference
+	back() const
+	{
+		return _data[size() - 1];
+	}
+
+	/**
+	 * Adds requested range to a transaction and returns slice.
+	 *
+	 * @param[in] start start index of requested range
+	 * @param[in] n number of elements in range
+	 * @return slice from start to start + n
+	 */
+	slice<pointer, reference>
+	range(size_type start, size_type n)
+	{
+		if (start + n >= N)
+			throw std::out_of_range("array::slice");
+
+		detail::conditional_add_range_to_tx(_data + start, n);
+
+		return {_data + start, _data + start + n};
+	}
+
+	/**
+	 * Returns const slice.
+	 *
+	 * @param[in] start start index of requested range
+	 * @param[in] n number of elements in range
+	 * @return slice from start to start + n
+	 */
+	slice<const_iterator, const_reference>
+	range(size_type start, size_type n) const
+	{
+		if (start + n >= N)
+			throw std::out_of_range("array::slice");
+
+		return {_data + start, _data + start + n};
+	}
+
+	/**
+	 * Returns const slice.
+	 *
+	 * @param[in] start start index of requested range
+	 * @param[in] n number of elements in range
+	 * @return slice from start to start + n
+	 */
+	slice<const_iterator, const_reference>
+	crange(size_type start, size_type n) const
+	{
+		if (start + n >= N)
+			throw std::out_of_range("array::slice");
+
+		return {_data + start, _data + start + n};
+	}
+
+	/**
+	 * Returns size of the array.
+	 */
+	constexpr size_type
+	size() const noexcept
+	{
+		return N;
+	}
+
+	/**
+	 * Returns the maximum size of the array.
+	 */
+	constexpr size_type
+	max_size() const noexcept
+	{
+		return N;
+	}
+
+	/**
+	 * Checks wheter array is empty.
+	 */
+	constexpr bool
+	empty() const noexcept
+	{
+		return size() != 0;
+	}
+
+	/**
+	 * Fills array with specified value
+	 * and adds entire array to a transaction.
+	 *
+	 * @throw transaction_error when adding the object to the
+	 *		transaction failed.
+	 */
+	void
+	fill(const_reference value)
+	{
+		detail::conditional_add_to_tx(this);
+		std::fill(_data, _data + size(), value);
+	}
+
+	/**
+	 * Swaps content with other array's content.
+	 * Adds both arrays to a transaction.
+	 *
+	 * @throw transaction_error when adding the object to the
+	 *		transaction failed.
+	 */
+	void
+	swap(array &other)
+	{
+		detail::conditional_add_to_tx(this);
+		detail::conditional_add_to_tx(&other);
+
+		std::swap_ranges(_data, _data + size(), other._data);
+	}
+
+private:
+	static constexpr bool small_array_snapshotting = N * sizeof(T) < 1024;
```

Comment:
make `1024` into a template parameter, and set its default value to
```
#ifndef
#define __PMEM_OBJ_ITERATOR_BULK_SNAPSHOT_SIZE 1024
#endif
```
or something like that.
This will allow users to customize it both globally (by simply defining the above value) and per-array.